File: src/remedy/connect/mdns.py
"""Same-LAN mDNS advertise for Grove Connect.

Service: ``_remedy-connect._udp`` on the RFC 6762 group ``224.0.0.251:5353``.
TXT carries a **host-pub hash** (blake2s-256 of the 32-byte public key, first
16 hex chars). Never the local API, Bearer token, or pair secret.

No extra dependency: packets are stdlib UDP. Unit tests use the encode helpers
and do not need a live multicast path.
"""

from __future__ import annotations

import contextlib
import hashlib
import logging
import socket
import struct
import threading
from collections.abc import Callable

from remedy.connect.bind import assert_chosen_bind, is_chosen_ipv4

log = logging.getLogger(__name__)

SERVICE_TYPE = "_remedy-connect._udp"
SERVICE_NAME = f"{SERVICE_TYPE}.local"
MDNS_GROUP = "224.0.0.251"
MDNS_PORT = 5353
DNS_PTR = 12
DNS_TXT = 16
DNS_A = 1
DNS_SRV = 33
DNS_IN = 1
# QR=1, AA=1 — unicast-looking mDNS announcement.
_MDNS_RESPONSE_FLAGS = 0x8400
_TTL = 120


def host_pub_hash(host_pub: bytes) -> str:
    """blake2s-256 of the host public key, truncated to 16 hex characters."""
    digest = hashlib.blake2s(bytes(host_pub), digest_size=32).hexdigest()
    return digest[:16]


def _advertise_ipv4(bind_host: str) -> str:
    host = assert_chosen_bind(bind_host)
    if not is_chosen_ipv4(host):
        raise ValueError(f"mDNS advertise needs a chosen IPv4, not {bind_host!r}")
    return host


def encode_dns_name(name: str) -> bytes:
    """Length-prefixed DNS labels, no compression."""
    out = bytearray()
    trimmed = name.strip().rstrip(".")
    if not trimmed:
        return b"\x00"
    for label in trimmed.split("."):
        raw = label.encode("ascii")
        if len(raw) > 63:
            raise ValueError(f"DNS label too long: {label!r}")
        out.append(len(raw))
        out.extend(raw)
    out.append(0)
    return bytes(out)


def encode_mdns_query(service: str = SERVICE_TYPE) -> bytes:
    """DNS-SD PTR question for the Grove Connect service (no secrets)."""
    qname = service if service.endswith(".local") else f"{service}.local"
    question = encode_dns_name(qname) + struct.pack("!HH", DNS_PTR, DNS_IN)
    header = struct.pack("!HHHHHH", 0, 0, 1, 0, 0, 0)
    return header + question


def _txt_rdata(entries: list[str]) -> bytes:
    out = bytearray()
    for item in entries:
        raw = item.encode("ascii")
        if len(raw) > 255:
            raise ValueError("TXT string too long")
        out.append(len(raw))
        out.extend(raw)
    return bytes(out)


def _rr(name: str, rtype: int, ttl: int, rdata: bytes) -> bytes:
    return encode_dns_name(name) + struct.pack("!HHIH", rtype, DNS_IN, ttl, len(rdata)) + rdata


def encode_mdns_announce(bind_host: str, port: int, host_pub: bytes) -> bytes:
    """Multicast DNS-SD announcement. TXT is host-pub hash only."""
    hp = host_pub_hash(host_pub)
    instance = f"Remedy-{hp}.{SERVICE_NAME}"
    target = f"remedy-{hp}.local"
    txt = _txt_rdata([f"hp={hp}"])
    srv = struct.pack("!HHH", 0, 0, int(port) & 0xFFFF) + encode_dns_name(target)
    a_rdata = socket.inet_aton(_advertise_ipv4(bind_host))
    answers = (
        _rr(SERVICE_NAME, DNS_PTR, _TTL, encode_dns_name(instance))
        + _rr(instance, DNS_SRV, _TTL, srv)
        + _rr(instance, DNS_TXT, _TTL, txt)
        + _rr(target, DNS_A, _TTL, a_rdata)
    )
    header = struct.pack("!HHHHHH", 0, _MDNS_RESPONSE_FLAGS, 0, 4, 0, 0)
    return header + answers


def start_advertiser(bind_host: str, port: int, host_pub: bytes) -> Callable[[], None]:
    """Advertise ``_remedy-connect._udp`` on the same L2. Returns a stop function.

    The packet never includes the API, Bearer, or pair secret — only a host-pub
    hash. Live send failures are swallowed so unit tests and locked-down NICs
    still get a working stop callback.
    """
    host = _advertise_ipv4(bind_host)
    packet = encode_mdns_announce(host, port, host_pub)
    stop = threading.Event()
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    with contextlib.suppress(OSError):
        sock.setsockopt(socket.IPPROTO_IP, socket.IP_MULTICAST_TTL, 255)
        sock.setsockopt(socket.IPPROTO_IP, socket.IP_MULTICAST_IF, socket.inet_aton(host))
    try:
        sock.bind((host, 0))
    except OSError:
        sock.close()

        def _already_stopped() -> None:
            return None

        return _already_stopped

    def _loop() -> None:
        while not stop.is_set():
            with contextlib.suppress(OSError):
                sock.sendto(packet, (MDNS_GROUP, MDNS_PORT))
            stop.wait(1.5)
        with contextlib.suppress(OSError):
            sock.close()

    thread = threading.Thread(target=_loop, name="connect-mdns", daemon=True)
    thread.start()
    log.info("Grove Connect mDNS advertising %s on %s:%s", SERVICE_TYPE, host, port)

    def _stop() -> None:
        stop.set()
        with contextlib.suppress(OSError):
            sock.close()
        thread.join(timeout=2.0)

    return _stop

File: native/go/connect/mdns.go
package connect

import (
	"encoding/binary"
	"encoding/hex"
	"errors"
	"fmt"
	"net"
	"strings"
	"sync"
	"time"
	"unicode"

	"golang.org/x/crypto/blake2s"
)

// Same-LAN mDNS advertise for Grove Connect.
//
// Service: _remedy-connect._udp on the RFC 6762 group 224.0.0.251:5353.
// TXT carries a host-pub hash (blake2s-256 of the public key, first 16 hex
// chars). Never the local API, Bearer token, or pair secret.
//
// No extra dependency beyond x/crypto (blake2s). Packets are stdlib UDP.

const (
	// ServiceType is the DNS-SD type for Grove Connect.
	ServiceType = "_remedy-connect._udp"
	// ServiceName is ServiceType under .local.
	ServiceName = ServiceType + ".local"
	// MDNSGroup is the IPv4 mDNS multicast address.
	MDNSGroup = "224.0.0.251"
	// MDNSPort is the mDNS UDP port.
	MDNSPort = 5353

	dnsPTR = 12
	dnsTXT = 16
	dnsA   = 1
	dnsSRV = 33
	dnsIN  = 1

	// QR=1, AA=1 — mDNS announcement response flags.
	mdnsResponseFlags = 0x8400
	mdnsTTL           = 120
	mdnsAnnounceEvery = 1500 * time.Millisecond
	mdnsStopJoin      = 2 * time.Second
)

// ErrMDNS is an mDNS encode / advertise policy failure (fail closed).
var ErrMDNS = errors.New("mdns error")

// HostPubHash is blake2s-256 of the host public key, truncated to 16 hex chars.
func HostPubHash(hostPub []byte) string {
	sum := blake2s.Sum256(hostPub)
	return hex.EncodeToString(sum[:])[:16]
}

func advertiseIPv4(bindHost string) (string, error) {
	host, err := AssertChosenBind(bindHost)
	if err != nil {
		return "", err
	}
	if !IsChosenIPv4(host) {
		return "", fmt.Errorf("%w: mDNS advertise needs a chosen IPv4, not %q", ErrMDNS, bindHost)
	}
	return host, nil
}

// EncodeDNSName encodes length-prefixed DNS labels with no compression.
func EncodeDNSName(name string) ([]byte, error) {
	trimmed := strings.TrimSpace(name)
	trimmed = strings.TrimRight(trimmed, ".")
	if trimmed == "" {
		return []byte{0}, nil
	}
	var out []byte
	for _, label := range strings.Split(trimmed, ".") {
		if len(label) > 63 {
			return nil, fmt.Errorf("%w: DNS label too long: %q", ErrMDNS, label)
		}
		for _, r := range label {
			if r > unicode.MaxASCII {
				return nil, fmt.Errorf("%w: DNS label not ASCII: %q", ErrMDNS, label)
			}
		}
		out = append(out, byte(len(label)))
		out = append(out, label...)
	}
	out = append(out, 0)
	return out, nil
}

// EncodeMDNSQuery builds a DNS-SD PTR question for the Grove Connect service.
// Empty service defaults to ServiceType. The packet carries no secrets.
func EncodeMDNSQuery(service string) ([]byte, error) {
	if service == "" {
		service = ServiceType
	}
	qname := service
	if !strings.HasSuffix(service, ".local") {
		qname = service + ".local"
	}
	name, err := EncodeDNSName(qname)
	if err != nil {
		return nil, err
	}
	question := make([]byte, 0, len(name)+4)
	question = append(question, name...)
	var rr [4]byte
	binary.BigEndian.PutUint16(rr[0:2], dnsPTR)
	binary.BigEndian.PutUint16(rr[2:4], dnsIN)
	question = append(question, rr[:]...)

	header := make([]byte, 12)
	binary.BigEndian.PutUint16(header[4:6], 1) // QDCOUNT
	return append(header, question...), nil
}

func txtRData(entries []string) ([]byte, error) {
	var out []byte
	for _, item := range entries {
		for _, r := range item {
			if r > unicode.MaxASCII {
				return nil, fmt.Errorf("%w: TXT string not ASCII", ErrMDNS)
			}
		}
		raw := []byte(item)
		if len(raw) > 255 {
			return nil, fmt.Errorf("%w: TXT string too long", ErrMDNS)
		}
		out = append(out, byte(len(raw)))
		out = append(out, raw...)
	}
	return out, nil
}

func resourceRecord(name string, rtype uint16, ttl uint32, rdata []byte) ([]byte, error) {
	enc, err := EncodeDNSName(name)
	if err != nil {
		return nil, err
	}
	hdr := make([]byte, 10)
	binary.BigEndian.PutUint16(hdr[0:2], rtype)
	binary.BigEndian.PutUint16(hdr[2:4], dnsIN)
	binary.BigEndian.PutUint32(hdr[4:8], ttl)
	binary.BigEndian.PutUint16(hdr[8:10], uint16(len(rdata)))
	out := make([]byte, 0, len(enc)+len(hdr)+len(rdata))
	out = append(out, enc...)
	out = append(out, hdr...)
	out = append(out, rdata...)
	return out, nil
}

// EncodeMDNSAnnounce builds a multicast DNS-SD announcement. TXT is host-pub hash only.
func EncodeMDNSAnnounce(bindHost string, port int, hostPub []byte) ([]byte, error) {
	hp := HostPubHash(hostPub)
	instance := "Remedy-" + hp + "." + ServiceName
	target := "remedy-" + hp + ".local"
	txt, err := txtRData([]string{"hp=" + hp})
	if err != nil {
		return nil, err
	}
	targetEnc, err := EncodeDNSName(target)
	if err != nil {
		return nil, err
	}
	srv := make([]byte, 6, 6+len(targetEnc))
	binary.BigEndian.PutUint16(srv[0:2], 0) // priority
	binary.BigEndian.PutUint16(srv[2:4], 0) // weight
	binary.BigEndian.PutUint16(srv[4:6], uint16(port&0xffff))
	srv = append(srv, targetEnc...)

	host, err := advertiseIPv4(bindHost)
	if err != nil {
		return nil, err
	}
	ip := net.ParseIP(host).To4()
	if ip == nil {
		return nil, fmt.Errorf("%w: mDNS advertise needs a chosen IPv4, not %q", ErrMDNS, bindHost)
	}

	instanceEnc, err := EncodeDNSName(instance)
	if err != nil {
		return nil, err
	}
	ptr, err := resourceRecord(ServiceName, dnsPTR, mdnsTTL, instanceEnc)
	if err != nil {
		return nil, err
	}
	srvRR, err := resourceRecord(instance, dnsSRV, mdnsTTL, srv)
	if err != nil {
		return nil, err
	}
	txtRR, err := resourceRecord(instance, dnsTXT, mdnsTTL, txt)
	if err != nil {
		return nil, err
	}
	aRR, err := resourceRecord(target, dnsA, mdnsTTL, []byte(ip))
	if err != nil {
		return nil, err
	}

	answers := append(append(append(ptr, srvRR...), txtRR...), aRR...)
	header := make([]byte, 12)
	binary.BigEndian.PutUint16(header[2:4], mdnsResponseFlags)
	binary.BigEndian.PutUint16(header[6:8], 4) // ANCOUNT
	return append(header, answers...), nil
}

// StartAdvertiser advertises _remedy-connect._udp on the same L2.
// Returns a stop function. Live send failures are swallowed so locked-down NICs
// still get a working stop callback. Bind failure fail-closes: no-op stop, nil error
// (server stays up; no fake multicast). Invalid bind host returns an error.
func StartAdvertiser(bindHost string, port int, hostPub []byte) (stop func(), err error) {
	host, err := advertiseIPv4(bindHost)
	if err != nil {
		return nil, err
	}
	packet, err := EncodeMDNSAnnounce(host, port, hostPub)
	if err != nil {
		return nil, err
	}

	conn, err := net.ListenUDP("udp4", &net.UDPAddr{IP: net.ParseIP(host), Port: 0})
	if err != nil {
		return func() {}, nil
	}

	done := make(chan struct{})
	var once sync.Once
	var wg sync.WaitGroup
	dst := &net.UDPAddr{IP: net.ParseIP(MDNSGroup), Port: MDNSPort}

	wg.Add(1)
	go func() {
		defer wg.Done()
		defer func() { _ = conn.Close() }()
		for {
			_, _ = conn.WriteToUDP(packet, dst)
			timer := time.NewTimer(mdnsAnnounceEvery)
			select {
			case <-done:
				timer.Stop()
				return
			case <-timer.C:
			}
		}
	}()

	stop = func() {
		once.Do(func() {
			close(done)
			_ = conn.Close()
		})
		wait := make(chan struct{})
		go func() {
			wg.Wait()
			close(wait)
		}()
		select {
		case <-wait:
		case <-time.After(mdnsStopJoin):
		}
	}
	return stop, nil
}
